Support for `wmin` and `wmax` parameters in Spectrum Factory

pkj-m opened the pull request:
<!-- This comments are hidden when you submit the pull request,
so you do not need to remove them! -->

<!-- Please be sure to check out our contributing guidelines,
https://github.com/radis/blob/develop/CONTRIBUTING.md . -->

### Description
<!-- Provide a general description of what your pull request does. -->

This PR adds support for new parameters `wmin` and `wmax` in the SpectrumFactory. Most modifications have been made in the `base.py` file at `radis/lbl`.
New function calls can now look like:

```
calc_spectrum(1200 * u.cm, 1400 * u.cm,...)
calc_spectrum(1200 * u.cm**-1, 1400 * u.cm**-1, ...)

calc_spectrum(1200, 1400, wunit = 'nm') 
calc_spectrum(1200, 1400, wunit = '1/m')
```
while the current function calls are still supported:

```
get_waverange(wavenum_min=1200, wavenum_max=1400)
get_waverange(wavelength_min=1200 * u.cm, wavelength_max=1400 * u.cm)
```
A few things that are still up for discussion are:

1. Can we allow users to pass units with both the `wmin`/`wmax` parameters *and* `wunit`? If so, which unit takes precedence if they differ?
2. Currently, the code thows an error if calls like ` calc_spectrum(1200, 1400, ...) are made. I would like to change that to an assertion error as I personally feel the function call looks way too vague and should not be allowed.

Will fix #56 after tests, docstring changes and further discussion.

pkj-m commented on radis/lbl/base.py:
```diff
@@ -3269,25 +3272,73 @@ def get_waverange(
 
     # Check input
+    if (
+        isinstance(wmin, u.Quantity) or isinstance(wmin, u.Quantity)
```
Fix `wmin` to `wmax`

erwanp commented:
Hello. I added some extra tests but I am not able to push them. Did you accept "Allow edits from Maintainers"? 

Anyway:

1. we will have the same problem when generalizing units to Spectrum methods #57 . Methods such as get() have a default "nm" argument. We don't want users to be able to use `100 / u.cm, wunit="nm"`. However, `wunit` is not `None` by default. I see the following solutions:

- **a)** `wunit` default becomes None, and we write in the Docstrings that if it is None and undimensioned units are given, we choose "nm" (for instance). > looks dangerous to me. 

- **b)** we allow dimensioned units to override `wunit`. > looks dangerous to me too. 
- **c)** i looked up [how to check whether optional functino parameter is set](https://stackoverflow.com/questions/14749328/how-to-check-whether-optional-function-parameter-is-set). The answer is no. The best idea i could find there was to create a `Default` class. For instance, we would have `Default("nm")` class. 
- **d)** alternatively, we can have our own default value, for instance a string that says "nm_by_default", and is replaced by "nm" at runtime if wmin is not dimensioned (no user confusion possible). 

Any preference? 

2. I agree that `calc_spectrum(wmin=1200, wmax=1400,...)` is vague, but it also quite efficient. I see two reasons why we cannot drop supporting this syntax at this point: 

- backward compatibility: for instance, the very first example of the documentation uses this syntax:
```
from radis import calc_spectrum
s = calc_spectrum(1900, 2300,         # cm-1
                  molecule='CO',
                  isotope='1,2,3',
                  pressure=1.01325,   # bar
                  Tgas=700,           # K
                  mole_fraction=0.1,
                  path_length=1,      # cm
                  )
s.apply_slit(0.5, 'nm')       # simulate an experimental slit
s.plot('radiance')
```
- Also, your remark also applies to all of the other arguments in the function above. The unit of pressure, or path_length, are not defined neither. We don't really want to force users to write them down all the time.

However, the additional uncertaintity for wmin, wmax is that not only there is a default unit, but also it can be wavelengths or wavenumbers. I see different ways to implement the defaults:
- **a.** keep `wavenum_min, wavenum_max` as the first two arguments, instead of `wmin, wmax`. So the historical syntax would work. And at the same time we force `wmin,wmax` to be dimensioned or `wunit` to be defined.
- **b.** `wmin, wmax` are the first two arguments, but `wunit` has a default value. There will be a problem if `wmin, wmax` are dimensioned and in contradiction with `wunit` default value. This is exactly the Problem 1. We will apply the same solution as discussed there.

pkj-m commented:
Hi @erwanp, sorry about that. I've checked the allow maintainer edits button now. Can you try pushing your code once again?

Regarding problem 1, I do like the clarity that option (d) provides without burdening the end-user with too many inputs. However, I am not sure how it can resolve the original problem, where both `wunit` and `wmin/wmax` have user-supplied contradicting units. 

It will however, resolve the second problem. Infact, for the second problem, where neither `wmin/wmax` nor `wunit` have a user-provided value, having our own `nm-by-default` kind of setup will be a neat way of handling this situation. 

I personally do not think having `wavenum_min` before `wmin` would make sense in the long run if our ultimate goal is to phase out these parameters and operate purely using `wmin/wmax`.

I can start working on implementing the `nm_by_default` setup for `wunit`, but we still need to decide on the right way to handle problem 1.

erwanp commented:
The push worked, thanks!

Option (d)  can solve 1. because now we can differenciate between cases where contradictory wunit was set by the user, and was the default:
```
calc_spectrum(wmin = 100*u.nm)   # where wunit is "cm-1" by default, but not a user input.
calc_spectrum(wmin = 100*u.nm, wunit="cm-1")  # where wunit is also "cm-1", but set that the user  
```
With (d), the first one will return wunit in "cm-1_by_default" and we know the user didn't mean anything wrong. So we can safely override it with the user input "u.nm".  
2nd line: something is definitly wrong, so we raise an (appropriate!) error. 

Is that ok for you for 1. ?  

---

(d) or (c) i'm not sure yet. (c) looks good too. The "Default" is never entered by the user, but appears in the fonction signature. 
For instance [calc_spectrum](https://radis.readthedocs.io/en/latest/source/radis.lbl.calc.html#radis.lbl.calc.calc_spectrum)  would look like: 

>  calc_spectrum(wmin=None, wmax=None, wunit=Default("cm-1"), ... )

Or 

>  calc_spectrum(wmin=None, wmax=None, wunit="cm-1_by_default")

Do you wanna ask on the Chat what's best? I can forward to my Lab people to have input. We can also make a Poll here. 


---

Ok for 2.b! So we give a default value to "wunit", with either option (c) or (d).

pkj-m commented:
Thanks for the elaborate explanation! I understand it perfectly now. So we can go ahead with using a default value for `wunit`. Between (c) and (d), I really don't see any particular advantage of either one, so I guess we should wait for a poll to decide how to move forward with that?

erwanp commented:
Do you wanna do the poll? I used [Github-Polls](https://github.com/apps/polls) in #63  if you want an example

pkj-m commented:
## Problem
After introducing new parameters `wmin`, `wmax` and `wunit` in the SpectrumFactory (which will eventually replace `wavelength_min/max` and `wavenum_min/max` parameters completely), we face a new problem of deciding the methods' behavior when the user enters `wmin/wmax` without any accompanying units *and* does not provide `wunit` either.

For instance, a call like `calc_spectrum(1200, 1400, ...)` will lead us to the above-described problem.

## Suggestion
To resolve the problem, we can specify a default value of `wunit` which will allow us to differentiate between any contradicting function calls made by the user (and raise an appropriate error) as well as handle the case where the user does not specify a unit either with `wmin/wmax` or with `wunit` easily.  The only thing left for further discussion is how do we implement the default value of the parameter. 

## Options

- Option 1: we create a default class, such as `Default`, and then set `wunit = Default("cm-1")` in the function definition.

`calc_spectrum(wmin=None, wmax=None, wunit=Default("cm-1"), ... )`

- Option 2: we hardcode a custom string, for instance, `nm_by_default` as the default value for `wunit`, which will again allow us to identify whether the value has been passed by the user (in which case it'll be `nm`, etc. to begin with), or it is the default value (in which case we internally convert it to `nm` before processing, and also realize that the units will now be supplied with the value of `wmin/wmax` directly.

`calc_spectrum(wmin=None, wmax=None, wunit="cm-1_by_default")`

I think one small advantage for option 1 might be the fact that having a `Default` class might help keep the code cleaner in case problems similar to this (with contradicting or absence of user-specified units) arise in the future.

## Votes

/polls "Option 1" "Option 2"

erwanp commented:
/polls "Default("cm-1")" ""cm-1_by_default""

erwanp commented:
/polls "Default("cm-1")" ""cm-1_by_default""

erwanp commented:
/polls "A" "B"

erwanp commented:
@pkj-m i couldn't get the /polls to work, so I replaced with 2 emojis. Asking for user input now.

pkj-m commented:
@erwanp Now that it seems all the votes have come in, should I start working towards option 1 or should we give it some more time?

erwanp commented:
Let's do! 
We should create a default class shared everywhere in Radis. That could be used to check default in other pull requests too #74 

There is a `Function arguments` part in `misc/utils.py`, maybe in there?

erwanp commented on radis/misc/utils.py:
```diff
@@ -96,6 +96,15 @@ def __del__(self):
 # ==============================================================================
 
 
+class Default:
+    """ Returns argument `unit` with an appended `_by_default` string
+    """
+
+    def __new__(cls, unit):
+        cls.unit = unit + "_by_default"
+        return cls.unit
+
+
```
In this version you'd still keep the "_by_default" string, which I find a bit verbosy.

What do you think of this:

```suggestion
class Default:
    """ Contains a value
    
    Examples
    --------
    
    ::
    
        a = Default("42")
        isinstance(a, Default)
        >>> True
        
        a.value
        >>> 42
    """

    def __init__(self, value):
        self.value= value

```

Which also has the advantage to work for anything (i.e. : `Default(1.mbar)` if we ever need it ) 

Could event implement the "__eq__" method so you can get "a == 42 >>> True" 

And then, when entering the function we can evaluate all the Default arguments as `a = a.value`

pkj-m replied:
Ah I see. The main reason why I implemented it with the `__new__` magic method was to avoid 
> when entering the function we can evaluate all the Default arguments as a = a.value

But I think you're right about the multi-functionality that your implementation has to offer! I'll change it right away.

pkj-m commented:
@erwanp I've introduced a new `Default` class according to the discussion we had, and also rewrote the `get_waverange` function to handle all the possible syntax that the user might use while calling the method.  Let me know if you think I should implement it differently.

Also, I had made a jupyter notebook consisting of all the different ways in which get_waverange function could be called to test my method. If you like, I can share the link to the notebook and we could use that to come up with new and more robust tests for this new implementation!

pkj-m commented:
@erwanp I have made all the changes you suggested. If this looks good to you, I can make the docstring changes. Also, let me know what you think about [these tests ](https://github.com/pkj-m/radis-lab/blob/master/Testing%20the%20new%20get_waverange%20method.ipynb) for this PR. If they look good to you then I can add them to the test suite as well.

erwanp commented:
Hello! I've looked at the test suite and they're very complete! 

To test the ones that should raise an error, you can use:
```
with pytest.raises(ValueError):
    # code that should raise ValueError  for this Test to pass
```

erwanp commented on radis/lbl/base.py:
```diff
@@ -3239,54 +3240,116 @@ def plot_populations(self, what="vib", isotope=None, nfig=None):
+    wavenum_present = wavenum_min is not None or wavenum_max is not None
+    wavelength_present = wavelength_min is not None or wavelength_max is not None
+    if w_present + wavenum_present + wavelength_present >= 2:
+        raise ValueError("Cannot pass more than one set of values as input")
```
```suggestion
        raise ValueError("Cannot pass more than one set of values as input: choose either wmin/wmax (with astropy.units), wavenum_min/wavenum_max, or wavelength_min/wavelength_max")
```

erwanp replied:
@pkj-m you can commit this directly btw.

erwanp commented on radis/lbl/base.py:
```diff
@@ -3239,54 +3240,116 @@ def plot_populations(self, what="vib", isotope=None, nfig=None):
+                assert not isinstance(wmax, u.Quantity)
+                wmin = wmin * u.Unit(wunit)
+                wmax = wmax * u.Unit(wunit)
+                represent_wavenum = True
```
it works, but are you sure we cannot simplify all these if/else?

pkj-m replied:
I think there must be a few simplifications we can make. I was under the impression that keeping the code structure like this would make it easier to follow and debug, especially now that we have quite a few ways of calling the `get_waverange` function.

erwanp replied:
We are starting with an `if` conditional on `wunit`, although it will be irrelevant if wmin, wmax are not given. 

In general, we are mixing (in the new implementation, but the old one too!) unit conversions to base units (nm / cm-1), final conversion to cm-1 (which requires to check the propagation medium: air or vacuum), and tests of input consistency (_anything with assertion tests, or raise errors_). It's a little hard to understand.

What about refactoring to check these sequentially:
```
# 1. Check input consistency
- Check that at least a pair of inputs is given 
- Check that not too types of inputs are given at the same time
- If wavenum or wavelengh, or wmin is dimensionned, check that wunit is consistent (if not default)
- [we can maybe leave the check that Medium is given in the wavelength case for later]

# 2. Conversion to base units (strip dimensioned units, convert to nm or cm )
# Now no need to assert / raise anything. 
wmin, wmax given
 --> if not dimensioned: convert to nm or cm-1 based on wunit  (no need to assert, we did it in 1.). Then write wavelength_min/max , or wavenum_min/max
 --> if dimensioned, convert to nm, or cm-1. Then write wavelength_min/max, or wavenum_min/max
wavenum_min, wavenum_max given
 --> convert to cm-1
wavelength_min, wavelength_max given:
--> convert to nm

# now we only have wavenum_min/max, or wavelen_min/max, in standard units (cm-1, or nm). Last thing to check: converting nm > cm-1 depends on the propagation medium (air/vacuum)
# 3. Convert to cm-1 
if wavenum_min/max: 
--> nothing to do 
if wavelength_min/max
--> check that medium is given
--> convert to cm-1 with nm2cm or nm_air2cm  (handy function in convert.py) 

Return wavenum_min/max
```

pkj-m replied:
Got it! I was a little confused and started thinking along the lines of simplifying the code from the logic point of view. I'll start working on this right away!

erwanp replied:
Any suggestion welcome, it's more of a design than a coding job ! :)

erwanp commented on radis/test/lbl/test_base.py:
```diff
@@ -400,12 +402,161 @@ def test_optically_thick_limit_2iso(verbose=True, plot=True, *args, **kwargs):
         radis.DEBUG_MODE = DEBUG_MODE
 
 
+def test_get_waverange(*args, **kwargs):
```
Great test suite!

I would also report here the comments of your [lab notebook](https://github.com/pkj-m/radis-lab/blob/master/lbl/Testing%20the%20new%20get_waverange%20method.ipynb), adding the "expected behavior". That will greatly improve the future readability of your test suite, if someone ever needs to update it:

Ex:
```
# wunit is none 
# ... wmin/wmax is none
# ...... wavenumber is passed  > should return the same wavenumbers

# ...... wavelength is passed > should convert and return wavenumbers

# ...... both wavelengths & wavenumbers are passed > should raise an Error

erwanp commented on radis/lbl/base.py:
```diff
@@ -3239,64 +3240,108 @@ def plot_populations(self, what="vib", isotope=None, nfig=None):
     wavelength_max=None,
     medium="air",
 ):
-    """ Returns wavenumber based on whatever input was given: either ν_min, ν_max 
-    directly, or λ_min, λ_max  in the given propagation ``medium``.
 
-    Parameters
-    ----------
-
-    medium: ``'air'``, ``'vacuum'``
-        propagation medium
-
-    wavenum_min, wavenum_max: float, or ~astropy.units.quantity.Quantity or ``None``
-        wavenumbers
-
-    wavelength_min, wavelength_max: float, or ~astropy.units.quantity.Quantity or ``None``
-        wavelengths in given ``medium``
-
-    Returns
-    -------
-
-    wavenum_min, wavenum_max,: float
-        wavenumbers
-
```
Keep the docstrings here 

(also add wmin, wmax, and wunit)

erwanp commented on radis/lbl/base.py:
```diff
@@ -3239,64 +3240,108 @@ def plot_populations(self, what="vib", isotope=None, nfig=None):
-
-    """
-
-    # Check input
```
Here, can you also keep a comment to explain what's happening ?  Like:
```
# 1. Checking consistency of all inputs 
```

erwanp commented on radis/lbl/base.py:
```diff
@@ -3239,64 +3240,108 @@ def plot_populations(self, what="vib", isotope=None, nfig=None):
+                    wavenum_min = wmin * u.Unit(wunit)
+                    wavenum_max = wmax * u.Unit(wunit)
+
+    if wavenum_min is not None or wavenum_max is not None:
```
Normally we already checked that if one is given the other one too. This can simply be:
```suggestion
    if wavenum_min:
```

erwanp commented on radis/lbl/base.py:
```diff
@@ -3239,64 +3240,108 @@ def plot_populations(self, what="vib", isotope=None, nfig=None):
+                else:
+                    wavenum_min = wmin * u.Unit(wunit)
+                    wavenum_max = wmax * u.Unit(wunit)
+
```
Comment needed:
```suggestion
    # We now have wavenum_min/max, or wavelength_min/max defined. Let's convert these to cm-1 (warning: propagating medium is required if we start from wavelengths!)
```

erwanp commented on radis/lbl/base.py:
```diff
@@ -3239,64 +3240,108 @@ def plot_populations(self, what="vib", isotope=None, nfig=None):
+        assert medium in ["air", "vacuum"]
+        wavelength_min = convert_and_strip_units(wavelength_min, u.nm)
+        wavelength_max = convert_and_strip_units(wavelength_max, u.nm)
         if medium == "air":
             wavelength_min_vac = air2vacuum(wavelength_min)
             wavelength_max_vac = air2vacuum(wavelength_max)
```
Or, if you find it nicer (no need to create this new _vac variables) :

```suggestion
        if medium == "air":
            wavenum_min = nm_air2cm(wavelength_max)
            wavenum_max = nm_air2cm(wavelength_min)
        else: # medium == 'vacuum':
            wavenum_min = nm2cm(wavelength_max)
            wavenum_max = nm2cm(wavelength_min)
```

pkj-m replied:
That definitely looks much cleaner! Thanks :D
